File: packages/core/src/scanner/GitDiff.ts
import { join } from 'node:path';

export function parseUnifiedDiff(
  repoPath: string,
  diffText: string
): Map<string, Set<number>> {
  const addedByFile = new Map<string, Set<number>>();
  let currentFile: string | null = null;
  let currentNewLine = 0;
  let inHunk = false;

  const hunkHeader = /^@@ -\d+(?:,\d+)? \+(\d+)(?:,(\d+))? @@.*/;

  for (const line of diffText.split(/\r?\n/)) {
    if (line.startsWith('diff --git ')) {
      currentFile = null;
      inHunk = false;
      continue;
    }
    if (line.startsWith('+++ ')) {
      const pathPart = line.replace(/^\+\+\+ /, '').trim();
      if (pathPart === '/dev/null') {
        currentFile = null;
      } else {
        const cleaned = pathPart.replace(/^b\//, '');
        currentFile = join(repoPath, cleaned);
        if (!addedByFile.has(currentFile)) {
          addedByFile.set(currentFile, new Set());
        }
      }
      continue;
    }
    if (line.startsWith('@@')) {
      const match = hunkHeader.exec(line);
      if (match) {
        const start = Number.parseInt(match[1] ?? '0', 10);
        currentNewLine = Number.isNaN(start) ? 0 : start;
        inHunk = true;
      } else {
        inHunk = false;
      }
      continue;
    }
    if (!inHunk || !currentFile) {
      continue;
    }
    if (line.startsWith('+') && !line.startsWith('+++')) {
      addedByFile.get(currentFile)?.add(currentNewLine);
      currentNewLine += 1;
      continue;
    }
    if (line.startsWith('-') && !line.startsWith('---')) {
      continue;
    }
    if (line.startsWith('\ No newline')) {
      continue;
    }
    currentNewLine += 1;
  }

  return addedByFile;
}

File: packages/intellij-plugin/src/main/kotlin/ai/astha/scantheplanet/idea/scanner/GitDiff.kt
package ai.astha.scantheplanet.idea.scanner

import java.nio.file.Path

object GitDiff {
    data class DiffHunk(
        val file: Path,
        val newStart: Int,
        val newCount: Int,
        val addedLines: Set<Int>
    )

    private val hunkHeader = Regex("^@@ -\\d+(?:,\\d+)? \\+(\\d+)(?:,(\\d+))? @@.*")

    fun parseUnifiedDiff(repoPath: Path, diffText: String): Map<Path, Set<Int>> {
        val addedByFile = mutableMapOf<Path, MutableSet<Int>>()
        var currentFile: Path? = null
        var currentNewLine = 0
        var inHunk = false

        for (line in diffText.lineSequence()) {
            when {
                line.startsWith("diff --git ") -> {
                    currentFile = null
                    inHunk = false
                }
                line.startsWith("+++ ") -> {
                    val pathPart = line.removePrefix("+++ ").trim()
                    if (pathPart == "/dev/null") {
                        currentFile = null
                    } else {
                        val cleaned = pathPart.removePrefix("b/")
                        currentFile = repoPath.resolve(cleaned).normalize()
                        addedByFile.putIfAbsent(currentFile, mutableSetOf())
                    }
                }
                line.startsWith("@@") -> {
                    val match = hunkHeader.matchEntire(line)
                    if (match != null) {
                        val start = match.groupValues[1].toIntOrNull() ?: 0
                        currentNewLine = start
                        inHunk = true
                    } else {
                        inHunk = false
                    }
                }
                inHunk && currentFile != null -> {
                    when {
                        line.startsWith("+") && !line.startsWith("+++") -> {
                            addedByFile[currentFile]?.add(currentNewLine)
                            currentNewLine += 1
                        }
                        line.startsWith("-") && !line.startsWith("---") -> {
                            // deletion: does not advance new line
                        }
                        line.startsWith("\\ No newline") -> {
                            // ignore
                        }
                        else -> {
                            currentNewLine += 1
                        }
                    }
                }
            }
        }
        return addedByFile
    }
}
